plasmapy/plasma/grids.py

```diff
@@ -76,6 +95,46 @@ def _validate(self):
     # Fundamental properties of the grid
     # *************************************************************************
 
+    def __repr__(self):
```

Comment:
Do you plan to have all sub-classes of `AbstractGrid` build and manage data around a `xarray.Dataset`?  If so, I would leverage `xarray.Dataset`'s `__repr__` because most have what you're doing here is already by the dataset's repr.

Reply:
Fair point, and i didn't think of that :(

Reply:
So, I tried doing that, but I think the dataset `__repr__` is a lot less informative than the one I created. Here's the `dataset.__repr__` for an example grid:

```
<xarray.Dataset>
Dimensions:               (ax0: 10, ax1: 10, ax2: 10)
Coordinates:
  * ax0                   (ax0) float64 -1.0 -0.7778 -0.5556 ... 0.7778 1.0
  * ax1                   (ax1) float64 -1.0 -0.7778 -0.5556 ... 0.7778 1.0
  * ax2                   (ax2) float64 -1.0 -0.7778 -0.5556 ... 0.7778 1.0
Data variables:
    B_x                   (ax0, ax1, ax2) float64 0.4924 -0.8735 ... -0.1033
    not_a_recognized_key  (ax0, ax1, ax2) float64 -0.04441 -0.2404 ... -0.3236
    B_y                   (ax0, ax1, ax2) float64 0.4924 -0.8735 ... -0.1033
    B_z                   (ax0, ax1, ax2) float64 0.4924 -0.8735 ... -0.1033
Attributes:
    axis_units:  [Unit("cm"), Unit("cm"), Unit("cm")]
``` 

And here's my `__repr__` for the same dataset:

```
*** Grid summary ***
<class 'plasmapy.plasma.grids.AbstractGrid'>
Shape: (10, 10, 10)
Units: [Unit("cm"), Unit("cm"), Unit("cm")]
Uniformly Spaced, dx,dy,dz = (0.222 cm,0.222 cm,0.222 cm)
-----------------------------
Recognized Quantities:
-> B_x (T)
-> B_y (T)
-> B_z (T)
-----------------------------
Unrecognized Quantities:
-> not_a_recognized_key (kg)
```

There's lots of additional information in the second one:
* Quantities separated by whether or not they were recognized, which is important for debugging
* Units for quantities
* The type of grid (in this case, AbstractGrid)
* Whether the grid is uniformly spaced, and the grid spacing.

So I think this printout is doing substantial work above the `dataset` one. Also, implementing this ourselves allows us to add more to it (eg. we could print the min and max of each axis and/or quantity).

Reply:
Per our discussion in the last meeting, I've fleshed out this `__repr__` a bit so that it now fully replaces the `xarray.dataset.__repr__`, while still including the additional features I wanted:

```
*** Grid Summary ***
<class 'plasmapy.plasma.grids.AbstractGrid'>
Dimensions: (ax0: 10, ax1: 10, ax2: 10)
Uniformly Spaced: (dax0, dax1, dax2) = (0.222 cm, 0.222 cm, 0.222 cm)
-----------------------------
Coordinates:
	-> ax0 (cm) float64 (10,)
	-> ax1 (cm) float64 (10,)
	-> ax2 (cm) float64 (10,)
-----------------------------
Recognized Quantities:
	-> B_x (T) float64 (10, 10, 10) 
	-> B_y (T) float64 (10, 10, 10) 
	-> B_z (T) float64 (10, 10, 10) 
-----------------------------
Unrecognized Quantities:
	-> not_a_recognized_key (kg) float64 (10, 10, 10) 
```